data/posts/remix-jam-2025-recap.md

```diff
@@ -0,0 +1,279 @@
+
+```tsx
+// books.tsx
+import type { RouteHandlers } from "@remix-run/fetch-router";
+import { Frame } from "@remix-run/dom";
+import { routes } from "../routes.ts";
+
+export default {
+  use: [loadAuth],
+  handlers: {
+    index() {
+      let books = getAllBooks();
+
+      return render(
+        <Layout>
+          <h1>Browse Books</h1>
+          {/* ... search and filter UI ... */}
+        </Layout>,
+      );
+    },
+
+    show({ params }) {
+      let book = getBookBySlug(params.slug);
+      if (!book) {
+        return render(<Layout>{/* ... 404 ... */}</Layout>, { status: 404 });
+      }
+
+      return render(
+        <Layout>{/* ... book details and add to cart form ... */}</Layout>,
+      );
+    },
+  },
+} satisfies RouteHandlers<typeof routes.books>;
+```
+
+Finally, Ryan joined Michael to tie it all together and showed two of the most exciting pieces:
+
+`hydrated()` and `Frame`.
+
+There's a lot more that needs to be said about these new APIs, but essentially:
+
+- `hydrated()` is the way to selectively hydrate a component to run JavaScript on the client-side (kind of like React's `"use client"`)
+- `Frame` is the primitive for async UI, heavily inspired by iframes (though not actually built on them)
+
+Ryan and Michael demoed how you can interop hydrated components inside of Frames and vice versa. And best of all, when you mutate some data and need to revalidate a frame, the response is raw HTML which our new, hybrid reconciler intelligently morphs into the existing DOM.
+
+It's a simple model with powerful primitives and technology, inspired by the best of the web and many other open source projects like React RSC, HTMX's [idiomorph](https://github.com/bigskysoftware/idiomorph), and much more.
+
+## What's Next
+
+![Remix Demo CDs in a display case with a hand reaching in to grab one](/blog-images/posts/remix-jam-2025-recap/remix-demo-cd.webp)
+
+You're probably wondering when we'll release Remix 3, what else we plan to include, and how to play with the code yourself.
+
+In terms of timeline, we don't have a hard commitment yet, but we want to get this to everyone in a stable state as soon as possible. Our goal right now is to continue shipping everything as small, composable modules (like we've been doing), and give you the first version of the full-stack framework package `remix` in early 2026.
+
+In the meantime:
+
+- [Watch the repo](https://github.com/remix-run/remix) (check out the bookstore demo)
+- Check out Remix team member Mark Dalgleish's [collection of Remix 3 resources](https://github.com/markdalgleish/remix3-resources) with demo apps, third-party integration examples, and timestamped links to every API shown at Remix Jam
+- [Subscribe to our newsletter](https://rmx.as/newsletter)
+- Keep reading our blog
+
+Hope to see many of you at Remix Jam 2026 💿
+
+[^1]: Long time community member @rossipedia put up [an improved version of the demo on GitHub](https://github.com/rossipedia/remix-jam-mk2/tree/main)
```

Comment:
We should link to the running version too: https://rossipedia.github.io/remix-jam-mk2/

There's also an even further improved version at https://github.com/rossipedia/remix-jam-mk3/ and https://rossipedia.github.io/remix-jam-mk3, but probably bnest to stick with the one closest to Ryan's original demo version.